### PR TITLE
Modernize original Train_models notebook: direct Ultralytics YOLO integration

### PR DESCRIPTION
**Pull request by louisrf**
- Replace kso_utils.train_models with direct Ultralytics YOLO API
- Comprehensive documentation with model size guidance
- Support YOLOv8/v9/v10/YOLO11 model families
- Add dataset verification with split checking and GPU detection
- Add test set evaluation phase for publishable metrics
- Remove W&B outputs. Saved to folders instead

### CODE REVIEW

File: notebooks/analyse/Train_models.ipynb [hunk 1]
```diff
@@ -2,284 +2,309 @@
...
+    "PROJECT_ID = \"\"          # Your LUMI project number\n",
+    "\n",
+    "# ── Paths ──\n",
+    "BASE_PATH = Path(f\"/scratch/project_{PROJECT_ID}/{USER}\")\n",
+    "data_path = BASE_PATH / \"datasets\" / \"\"        # Dataset folder containing data.yaml\n",
+    "EXPERIMENT_ROOT = BASE_PATH / \"models\"         # Where training runs are saved\n",
+    "\n",
+    "# ── Experiment ──\n",
+    "exp_name = \"\"                    # Name for this run (creates a subfolder)\n",
+    "baseline_weights = \"yolo11m.pt\"  # Ultralytics model name OR path to your .pt weights\n",
```
[the remark this answers is not in the document]
Comment (louisrf) [reply notebooks/analyse/Train_models.ipynb:78]: YOLO11 is a curreny standard Ultralytics model

**Review by trossi — CHANGES_REQUESTED** [review]
The notebook is clean and easy to follow. Some comments:

---

Similarly to #502, I'd find the first cell simpler (and more flexible) if the user gives directly the required paths:
```py
data_path = Path("<yolo_data_path>")
EXPERIMENT_ROOT = Path("<models_directory_path>")
```

---

To get the notebook working without wandb, I needed to add the following lines before training:
```py
from ultralytics import settings
settings.update({"wandb": False})
```

---

The YOLO runs seem to use unsuitable number of workers by default (pytorch gives a warning), so it'd be best to set the number of workers to match the number of requested CPU cores. We could do this for example in the Phase 2 cell by adding:
```py
# Set number of workers to the number of requested CPU cores
NUM_WORKERS = len(os.sched_getaffinity(os.getpid()))//2
...
print(f\"Number of workers: {NUM_WORKERS}\")\n"
```
and then use this variable in YOLO calls:
```py
model.train(..., workers=NUM_WORKERS, ...)
```
and
```py
model.eval(..., workers=NUM_WORKERS, ...)
```

---

If the notebook is run multiple times with the same experiment name, the new data gets appended to `results.csv` and the plots become messy. Maybe there could be a check in the verification phase that the experiment directory is empty?

**Comment by louisrf** [comment]
@trossi great advice, I'm going to implement some changes tonight or tomorrow and send it through

- new config is definitely better
- I'll make sure we have exp_name → exp_name2 → ...) back
- also passing workers=NUM_WORKERS to both model.train() and model.val()
- For W&B disable I've been using import os
os.environ["WANDB_MODE"] = "disabled" 

in the train phase before importing YOLO, but I'll test/integrate 
 
from ultralytics import settings
settings.update({"wandb": False})

in Phase 2 if that works, even better.

**Comment by trossi** [comment]
@louisrf Great!

> number of workers

A small update: the number of workers above might not work correctly outside LUMI, so here is a bit more robust one to ensure that there is at least one worker:
```py
NUM_WORKERS = max(1, len(os.sched_getaffinity(os.getpid()))//2)
```

> I'll make sure we have exp_name → exp_name2 → ...) back

I found it confusing in the old notebook that there was an automatic incrementing of the experiment name, resulting in that the experiment name variable didn't actually correspond to the real experiment name. Alternative could be to raise an error in the Phase 2 if the experiment directory is not empty, so that the user needs to change the name explicitly?

**Comment by louisrf** [comment]
@trossi Just to make sure we're on the same page:

User first sets exp_name = "cod_v1"
First run creates models/cod_v1/ 
User reruns/restarts kernel → Phase 2 after config raises an error: "directory not empty, change exp_name"
User changes to exp_name = "cod_v2" and reruns 
This has always kind of made more sense to me. It makes sure exp_name always matches the actual folder, this way you always give each experiment a meaningful name

**Comment by trossi** [comment]
> It makes sure exp_name always matches the actual folder, this way you always give each experiment a meaningful name

This sounds logical to me too. I'm thinking though that should this check be done in the training cell? Then, user would be able to run evaluation cell for old models (skipping training).

**Comment by louisrf** [comment]
@trossi added the exp directory check in Phase 3, and tested that, the logic works well! Committed the updated notebook, I'll monitor checks, run and formatting

**Review by trossi — APPROVED** [review]

**Comment by louisrf** [comment]
Content incorporated into #506 (reorganize-notebooks). Closed this PR. Thanks @trossi for the review. Feedback is applied in #506.
